File: Data Structures & Algorithms/is-palindrome/submission-4.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isPalindrome(s) {
        s = s.split("")
        s = s.filter(char => /^[a-z0-9]*$/gi.test(char))

        let left = 0
        let right = s.length - 1

        while (left <= right){
            if (s[left].toLowerCase() !== s[right].toLowerCase()){
                return false
            }
            else{
                left += 1
                right -= 1
            }
        }

        return true

    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        #where the split occurs between two nodes is always the LCA
        #we can use recursive dfs to find the split

        cur = root

        while cur:
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left

            #otherwise the cur is splitting the two nodes, which means it is the LCA and we can return it

            else:
                return cur
            
            

        